server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -617,6 +635,31 @@ public Object call() throws IOException
     // NB: The rows are still in memory until they're done persisting, but we only count rows in active indexes.
     rowsCurrentlyInMemory.addAndGet(-numPersistedRows);
     bytesCurrentlyInMemory.addAndGet(-bytesPersisted);
+
+    log.info("Persisted rows[%,d] and bytes[%,d]", numPersistedRows, bytesPersisted);
+
+    // bytesCurrentlyInMemory can change while persisting due to concurrent ingestion.
+    // Hence, we use bytesInMemoryBeforePersist to determine the change of this persist
+    if (!skipBytesInMemoryOverheadCheck && bytesInMemoryBeforePersist - bytesPersisted > maxBytesTuningConfig) {
+      // We are still over maxBytesTuningConfig even after persisting.
+      // This means that we ran out of all available memory to ingest (due to overheads created as part of ingestion)
+      String errorMsg = StringUtils.format("Failing task as task uses up too much heap memory. "
+                                           + "Persist can no longer free up memory. Objects that "
+                                           + "cannot be freed up from intermediate persist include Sinks, Memory Mapped Hydrants, "
+                                           + "and other overhead created while ingesting/pervious persistings. "
+                                           + "This check, along with calculation and consideration of overhead objects' "
+                                           + "memory can be disabled by setting skipBytesInMemoryOverheadCheck to true "
+                                           + "in the tuningConfig of the ingestionSpec "
+                                           + "(note that doing so can result in OOME). "
+                                           + "numSinks[%d] numHydrantsAcrossAllSinks[%d] totalRows[%d]",
+                                           sinks.size(),
+                                           sinks.values().stream().mapToInt(Iterables::size).sum(),
+                                           getTotalRowCount());
+      log.makeAlert(errorMsg)
+         .addData("dataSource", schema.getDataSource())
+         .emit();
+      throw new RuntimeException(errorMsg);
```

Comment:
ok, I walk back my previous suggestion about consolidating alert and exception messages, this is maybe a bit wordy for an alert, but I like including the expanded description in the task log. How about making the alert message a subset of the exception message, something like this: 

```suggestion
final String alertMessage = StringUtils.format(
          "Task has exceeded safe estimated heap usage limits, failing "
          + "(numSinks: [%d] numHydrantsAcrossAllSinks: [%d] totalRows: [%d])",
          sinks.size(),
          sinks.values().stream().mapToInt(Iterables::size).sum(),
          getTotalRowCount()
      );
      final String errorMessage = StringUtils.format(
          "%s.\nThis can occur when the overhead from too many intermediary segment persists becomes to "
          + "great to have enough space to process additional input rows. This check, along with metering the overhead "
          + "of these objects to factor into the 'maxBytesInMemory' computation, can be disabled by setting "
          + "'skipBytesInMemoryOverheadCheck' to 'true' or increasing 'maxBytesInMemory' (note that doing so might "
          + "allow the task to naturally encounter a 'java.lang.OutOfMemoryError').",
          alertMessage
      );
      log.makeAlert(alertMessage)
         .addData("dataSource", schema.getDataSource())
         .emit();
      throw new RuntimeException(errorMessage);
```

Reply:
LGTM. Reworded a little since increasing 'maxBytesInMemory' does not disabled the new check.